### ISSUE TITLE
Replace raw pointer with `std::reference_wrapper` where applicable

### ISSUE THREAD

**Issue by elBoberido** [eclipse-iceoryx/iceoryx#1071]
## Brief feature description

Currently we use raw pointer at several locations in our code base. This requires us to do `nullptr` checks even when we know that it is not possible to get a `nullptr` when the API is used in the correct way. Just returning a reference and expecting a reference as argument is not always possible or might be the source for different errors. By using `std::reference_wrapper` or an own implementation if the STL type does not give us the desired behavior, we can communicate that we do not expect `nullptr` and make our code base more robust towards a wrong usage.

## Detailed information

A problem with returning plain references is a potential unintentional copy on the caller side, e.g.
```cpp
Foo& getFoo() {
   return fooFromShm();
}

void releaseFoo(Foo& foo) {
    // code
}

int main() {
    auto foo = getFoo(); // copy occurs here
    releaseFoo(foo);  // release a stack based object
}
```
To prevent this, a `std::reference_wrapper` might be returned instead of a plain reference.

Places where this could be used are the `SubscriberPort`, `PublisherPort`, `ClientPort`, `ServerPort`, and probably many more places.

### To Do
- [ ] remove `cxx::not_null`

### Related to
[#717 Replace cxx::not_null with reference wherever possible](https://github.com/eclipse-iceoryx/iceoryx/issues/717)
[#1101 Return user header ptr instead of reference](https://github.com/eclipse-iceoryx/iceoryx/issues/1101)

**Comment by elfenpiff** [comment]
I would go even further that we try to remove raw pointer from the whole code base with very few exceptions like allocators or where we send relative pointers to shared memory constructs (but even there we do not use raw pointers).

* First of all we optimize the `cxx::optional` so that when a pointer is stored we indicate via `nullptr` if its empty or not and remove the additional bool.
* We replace all arguments which can be `nullptr` with that optional.
* `void*` can be replaced by a `struct void_ptr` or something like this, which behaves like a `void*` but is typed.

**Comment by elBoberido** [comment]
I realized that some time ago I played around with something called `SafeRef` with ref counting for all references which arehandled out. When the object which is referenced is destroyed, the error handler would be called when there are still references around. This would be only for debug builds and for release builds it would be a simple reference wrapper

**Comment by elBoberido** [comment]
cc @MatthiasKillat

**Comment by elBoberido** [comment]
@elfenpiff it's funny that you added `remove 'not_null'` to the issue description because this might be the solution to our problem. I tried to use `std::reference_wrapper` with the `cxx::optional` but it didn't work out of the box. It's either an issue with the `cxx::optional` or the `cxx::function_ref`. I didn't investigate it further.

@mossmaurice do you know if `cxx::function_ref` can handle a `std::reference_wrapper` as parameter?

The `cxx::not_null` might be a good solution but not with the current implementation. We need something more sophisticated and even though it comes from the core guidelines, I don't like the name and that one has to specify the type as pointer, i.e. `not_null<T*>` instead of `not_null<T>`. I think it is also poor to describe what it is not instead of what it is. One could think about a `some<T>`. Compare these two
```cpp
optional<not_null<ChunkHeader*>> takeChunk();

optional<some<ChunkHeader>> takeChunk(); 
```
Okay, it's inspired by the Rust `Option::Some<T>` and might be a bit alien to C++ developers. Unfortunately `std::ref` is already used to create a `std::reference_wrapper` and having a `cxx::ref` with the semantics I described above might confuse the C++ developer even more. I think this is something we can discuss after the v2 release in detail.

cc @MatthiasKillat @dkroenke @FerdinandSpitzschnueffler

**Comment by mossmaurice** [comment]
> do you know if `cxx::function_ref` can handle a `std::reference_wrapper` as parameter?

Yes, if this is what you mean:

```
class Functor
{
  public:
    int operator()()
    {
        return 42;
    }
};
Functor myFunctor;
std::reference_wrapper<Functor> myFunctorRef{myFunctor};
function_ref<int()> sut(myFunctorRef);
```

We should also compare our `function_ref` implementation against the latest [proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p0792r6.html) from 2022/01/17.

**Comment by elBoberido** [comment]
@mossmaurice no, I meant if something like `cxx::function_ref<Foo&(std::reference_wrapper<Bar>)>` works?
